### PR TITLE
[VarDumper] Add and use Caster::PREFIX_* consts

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Will need a rebase on top of #14058 once it is merged

### CODE REVIEW

File: src/Symfony/Component/VarDumper/Caster/DOMCaster.php [hunk 1]
```diff
@@ -154,8 +155,10 @@ public static function castDocument(\DOMDocument $dom, array $a, Stub $stub, $is
             'preserveWhiteSpace' => $dom->preserveWhiteSpace,
             'recover' => $dom->recover,
             'substituteEntities' => $dom->substituteEntities,
-            "\0~\0xml" => $dom->saveXML(),
         );
+        if (!($filter & Caster::EXCLUDE_VERBOSE)) {
+            $a += array(Caster::PREFIX_VIRTUAL.'xml' => $dom->saveXML());
+        }
 
         $dom->formatOutput = $formatOutput;
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Caster/DOMCaster.php]: I suggest moving the changing of ``formatOutput`` and its resetting to the previous value inside the ``if (!($filter & Caster::EXCLUDE_VERBOSE)) {`` check as it is needed only for the ``saveXML`` call AFAIK

Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DOMCaster.php]: good idea, done

File: src/Symfony/Component/VarDumper/Caster/Caster.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+            switch (true) {
+                case (self::EXCLUDE_NULL & $filter) && null === $v:
+                case (self::EXCLUDE_EMPTY & $filter) && empty($v):
+                case (self::EXCLUDE_NOT_IMPORTANT & $filter) && !in_array($k, $listedProperties, true):
+                    $type = $filter;
+                    break;
+                case isset($k[1]) && "\0" === $k[0]:
+                    switch ($k[1]) {
+                        case '~': $type = self::EXCLUDE_VIRTUAL; break;
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Caster/Caster.php]: I would make this multiline to make it more readable (seeing easily that we have some login in the case, not only a condition falling through to the next case).
currently, the code here is very similar to the code on lines 85 and 86 at first look, while the structure of the code is actually totally different

Comment (stof) [inline src/Symfony/Component/VarDumper/Caster/Caster.php]: I would make this multiline to make it more readable (seeing easily that we have some login in the case, not only a condition falling through to the next case).
currently, the code here is very similar to the code on lines 85 and 86 at first look, while the structure of the code is actually totally different

**Comment by nicolas-grekas** [comment]
This now need a rebase on top of #14114 before being merged

**Comment by nicolas-grekas** [comment]
Tests are green, ready for a merge, ping @symfony/deciders

**Comment by stof** [comment]
:+1:

File: src/Symfony/Component/VarDumper/Caster/AmqpCaster.php [hunk 3]
```diff
@@ -46,7 +46,7 @@ class AmqpCaster
 
     public static function castConnection(\AMQPConnection $c, array $a, Stub $stub, $isNested)
     {
-        $prefix = "\0~\0";
+        $p = Caster::PREFIX_VIRTUAL;
```
Comment (xabbuh) [inline src/Symfony/Component/VarDumper/Caster/AmqpCaster.php]: Can't we keep the old variable name? To me, `$prefix` is more meaningful than a one-character variable name.

Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/AmqpCaster.php]: Yes we can, done in fact

**Comment by xabbuh** [comment]
:+1:

**Pull request merged**